subiquity/models/keyboard.py

```diff
@@ -0,0 +1,124 @@
+
+from collections import defaultdict
+import gzip
+import io
+import logging
+import os
+import re
+
+from subiquitycore.utils import run_command
+
+log = logging.getLogger("subiquity.models.keyboard")
+
+etc_default_keyboard_template = """\
+# KEYBOARD CONFIGURATION FILE
+
+# Consult the keyboard(5) manual page.
+
+XKBMODEL="pc105"
+XKBLAYOUT="{layout}"
+XKBVARIANT="{variant}"
+XKBOPTIONS=""
+
+BACKSPACE="guess"
+"""
+
+class KeyboardModel:
+    def __init__(self, root):
+        self.root = root
+        self.layout = 'us'
+        self.variant = ''
+        self._kbnames_file = os.path.join(os.environ.get("SNAP", '.'), 'kbdnames.gz')
+        self._clear()
+        if os.path.exists(self.config_path):
+            content = open(self.config_path).read()
+            pat_tmpl = '(?m)^\s*%s=(.*)$'
+            log.debug("%r", content)
+            layout_match = re.search(pat_tmpl%("XKBLAYOUT",), content)
+            if layout_match:
+                log.debug("%s", layout_match)
+                self.layout = layout_match.group(1).strip('"')
+            variant_match = re.search(pat_tmpl%("XKBVARIANT",), content)
+            if variant_match:
+                log.debug("%s", variant_match)
+                self.variant = variant_match.group(1).strip('"')
+                if self.variant == '':
+                    self.variant = None
+
+    @property
+    def config_path(self):
+        return os.path.join(self.root, 'etc', 'default', 'keyboard')
+
+    @property
+    def config_content(self):
+        return etc_default_keyboard_template.format(layout=self.layout, variant=self.variant)
+
+    def has_language(self, code):
+        self.load_language(code)
+        return bool(self.layouts)
+
+    def load_language(self, code):
+        if code == self.current_lang:
+            return
+
+        self._clear()
+
+        raw = gzip.open(self._kbnames_file)
+        try:
+            with io.TextIOWrapper(raw) as kbdnames:
```

Comment:
Does kbdnames need to be compressed? It would be on a CD, in a squashfs that is already compressed, etc.